server/controllers/login.js

```diff
@@ -0,0 +1,131 @@
+        if (!scope.includes('read:org') || !scope.includes('user:email')){
+            return res.status(401).json({
+                message: 'More permissions are required'
+            })
+        }
+
+        //Fetch user email from github
+        const ghUserEmailResponse = await axios({
+            method: 'get',
+            url: `${constants.githubAPIDomain}/user/emails`,
+            headers: {
+                accept: 'application/vnd.github.v3+json',
+                Authorization: `token ${ghTokenResponse.data.access_token}`
+            }  
+        })
+
+        //Check if the user exists in our db
+        let user = await User.findOne({email: ghUserEmailResponse.data[0].email})
+        if(!user){
+            //Create a new user if the user does not exist in our db
+            const newUser = {
+                'email': ghUserEmailResponse.data[0].email
+            }
+            //Fetch user info from github
+            const ghUserResponse = await axios({
+                method: 'get',
+                url: `${constants.githubAPIDomain}/user`,
+                headers: {
+                    accept: 'application/vnd.github.v3+json',
+                    Authorization: `token ${ghTokenResponse.data.access_token}`
+                }  
+            })
+            newUser['name'] = ghUserResponse.data.name || ghUserResponse.data.login
+            newUser['username'] = ghUserResponse.data.login
+            newUser['avatarUrl'] = ghUserResponse.data.avatar_url
+
+            /*
+                In order to handle a case when user tries signing in gets created on RC but fails get stored in our db due to some error
+                To keep it consistent with RC we store the (user-email,encrypted-rcPassword) pair in our fs so that it can be used
+                to store later in our db
+            */
+            let inconsistentUsers = jsonfile.readFileSync(inconsistentUsersPath)
+            let rcPassword = inconsistentUsers[newUser.email]
+            
+            if(rcPassword){
+                //If the user already exists on RC
+                delete inconsistentUsers[newUser.email]
+                jsonfile.writeFileSync(inconsistentUsersPath, inconsistentUsers)
+            } else {
+                //Generate RC password and create user on RC
+                rcPassword = Math.random().toString(36).slice(2) +  Math.random().toString(36).toUpperCase().slice(2)
+                await axios({
+                    method: 'post',
+                    url: `${constants.rocketChatAPIURL}/users.register`,
+                    headers: {
+                        'Content-type': 'application/json'
+                    },
+                    data: `{
+                        "name": "${newUser.name}",
+                        "email": "${newUser.username}@rc4git.com",
+                        "pass": "${rcPassword}",
+                        "username": "${newUser.username}_rc4git"
```

Comment:
What if the username is same for `github` and `gitlab`? We need to handle this case too since we are supporting login from multiple platforms.